back/spec/models/basket_spec.rb

```diff
@@ -68,4 +68,18 @@
       expect(basket_idea).to be_invalid
     end
   end
+
+  context "when the basket's project is updated to non-budgeting participation method" do
+    let(:project) { create(:continuous_budgeting_project, min_budget: 200) }
+    let(:idea) { create(:idea, budget: 100, project: project) }
+
+    # Check the basket remains valid and thus won't fail data consistency checks, as would be the case,
+    # for example, if we enforce validation that the participation_context is budgeting.
+    it "the basket remains valid" do
+      basket = create(:basket, ideas: [idea], participation_context: project, submitted_at: Time.now)
```

Comment:
I created the basket inside the `it` block, as when the `:in_budgeting_participation_context` is enabled (not removed) this errors nicely, with:
```
1) Basket when the basket's project is updated to non-budgeting participation method the basket remains valid
     Failure/Error: expect(basket).to be_valid
       expected #<Basket id: "af44d31e-a0b2-4f5a-a28d-71ce98e60c39", submitted_at: "2021-08-30 17:34:21.797314000 +0000", user_id: "449efe15-5195-437e-bd1d-7c8b5cd69613", participation_context_id: "ebdc5407-72a1-4a4f-99be-3151d17dc27f", participation_context_type: "Project", created_at: "2021-08-30 17:34:21.970694000 +0000", updated_at: "2021-08-30 17:34:21.970694000 +0000"> to be valid, but got errors: Participation context is not a in budgeting method
```
Whereas when I declare the basket with `let(:basket) { create(:basket, ideas: [idea], participation_context: project, submitted_at: Time.now) }` _before_ the `it` block, the error occurs before the `expect` assertion:
```
Failures:

  1) Basket when the basket's project is updated to non-budgeting participation method the basket remains valid
     Failure/Error: let(:basket) { create(:basket, ideas: [idea], participation_context: project, submitted_at: Time.now) }
     
     ActiveRecord::RecordInvalid:
       Validation failed: Participation context is not a in budgeting method
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/evaluation.rb:18:in `create'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:12:in `block in result'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:9:in `tap'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy/create.rb:9:in `result'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory.rb:43:in `run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory_runner.rb:29:in `block in run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/factory_runner.rb:28:in `run'
     # /usr/local/bundle/gems/factory_bot-6.2.0/lib/factory_bot/strategy_syntax_method_registrar.rb:28:in `block in define_singular_strategy_method'
     # ./spec/models/basket_spec.rb:73:in `block (3 levels) in <top (required)>'
     # ./spec/models/basket_spec.rb:82:in `block (3 levels) in <top (required)>'
     # /usr/local/bundle/gems/webmock-3.12.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

Top 1 slowest examples (0.65629 seconds, 22.5% of total time):
  Basket when the basket's project is updated to non-budgeting participation method the basket remains valid
    0.65629 seconds ./spec/models/basket_spec.rb:79
```

Reply:
Acceptable to me. If I understand correctly, the issue you're preventing is that you don't want the validation errors to be thrown when doing `create(:basket)`? In that case, a trick that typically works is to not persist the instance yet in the database. You can do this with `build(:basket)`. Then validations will only run for the first time when doing `basked.valid?`, which is I think what you want. The disadvantage however, is that you wont be able to create proper relationships with other instances (as this requires the basket to be persisted in the database).

Reply:
Ah no, I understand the issue now. The problem is that `let` is lazy in evaluation. This means the basket will only be created when the line `basket.reload` is evaluated, which is too late and therefore throws the error. You can use `let!` instead and you will see that it then works (the basket will be created before the project update). But it's still find to also create the basket within the `it` block :)

Reply:
@sebastienhoorens Ah, that's a great explanation. :+1: Thanks. I'll think about which I prefer and merge when I decide (soon).